src/psyclone/psyir/transformations/loop_swap_trans.py

```diff
@@ -0,0 +1,159 @@
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors R. W. Ford, A. R. Porter, and S. Siso STFC Daresbury Lab
+#         A. B. G. Chalk STFC Daresbury Lab
+#         J. Henrichs, Bureau of Meteorology
+# Modified I. Kavcic, Met Office
+
+''' This module provides the loop swap transformation.'''
+
+from psyclone.psyir.nodes import Schedule
+from psyclone.psyir.transformations import LoopTrans, TransformationError
+
+
+class LoopSwapTrans(LoopTrans):
+    ''' Provides a loop-swap transformation, e.g.:
+
+    .. code-block:: fortran
+
+        DO j=1, m
+            DO i=1, n
+
+    becomes:
+
+    .. code-block:: fortran
+
+        DO i=1, n
+            DO j=1, m
+
+    This transform is used as follows:
+
+     >>> from psyclone.parse.algorithm import parse
+     >>> from psyclone.psyGen import PSyFactory
+     >>> ast, invokeInfo = parse("shallow_alg.f90")
+     >>> psy = PSyFactory("gocean1.0").create(invokeInfo)
+     >>> schedule = psy.invokes.get('invoke_0').schedule
+     >>> # Uncomment the following line to see a text view of the schedule
+     >>> # schedule.view()
+     >>>
+     >>> from psyclone.transformations import LoopSwapTrans
+     >>> swap = LoopSwapTrans()
+     >>> swap.apply(schedule.children[0])
+     >>> # Uncomment the following line to see a text view of the schedule
+     >>> # schedule.view()
+
+    '''
+    def __str__(self):
+        return "Exchange the order of two nested loops: inner becomes " + \
+               "outer and vice versa"
+
+    def validate(self, node_outer, options=None):
+        # pylint: disable=arguments-differ
```

Comment:
One more check, could you also prevent to swap loops with Calls and CodeBloks inside (anywhere start/stop/step/body), this is because this may have side-effects and we are changing the order. e.g:

```
do i = 1, 10
  do j = 1, 10
     write_to_file(i, j)
     write(*,*) i , j  ! Codeblock
  enddo
enddo
```
Swapping would lead to a different outputs. (or worse deadlocks if it were MPI send/recvs )

I think is just a matter of adding them in the `excluded_node_types` class attribute and the super will deal with the check.

(Hopefully this is not a problem for any existing test, in the APIs kernel calls don't inherit from Call and we must guarantee they are side-effect free, so this shouldn't impact PSy-layer code)

Reply:
Done - and all other tests seem to be fine with that.